File: python/tests/test_basic.py
from unicode_width_approximation import (get_codepoint_width, get_string_width,
                                         is_wide_char, is_zero_width)


def test_ascii_strings():
    assert get_string_width("") == 0
    assert get_string_width("Hello") == 5
    assert get_string_width("Hello, World!") == 13
    assert get_string_width("abc123") == 6


def test_cjk_strings():
    assert get_string_width("中文") == 4
    assert get_string_width("日本語") == 6
    assert get_string_width("한국어") == 6


def test_mixed_strings():
    assert get_string_width("Hello, 世界") == 11


def test_emoji_strings():
    assert get_string_width("🐱") == 2
    assert get_string_width("🙈🙉🙊") == 6
    assert get_string_width("👋🏻") == 2
    assert get_string_width("👨‍👩‍👧‍👦") == 2
    assert get_string_width("🇨🇳") == 2


def test_combining_marks():
    assert get_string_width("é") == 1
    assert get_string_width("ä") == 1
    assert get_string_width("é̂") == 1


def test_control_characters():
    assert get_string_width("hello\n") == 5
    assert get_string_width("hello\tworld") == 10


def test_fullwidth_characters():
    assert get_string_width("ＡＢＣ") == 6


def test_codepoint_width():
    # ASCII
    assert get_codepoint_width(ord("A")) == 1
    assert get_codepoint_width(ord("z")) == 1
    assert get_codepoint_width(ord(" ")) == 1

    # Control characters
    assert get_codepoint_width(0x00) == 0  # NULL
    assert get_codepoint_width(0x0A) == 0  # LF
    assert get_codepoint_width(0x0D) == 0  # CR

    # CJK
    assert get_codepoint_width(0x4E00) == 2
    assert get_codepoint_width(0x3042) == 2  # Hiragana A

    # Fullwidth
    assert get_codepoint_width(0xFF01) == 2  # Fullwidth !
    assert get_codepoint_width(0xFF21) == 2  # Fullwidth A

    # Emoji
    assert get_codepoint_width(0x1F600) == 2  # Grinning face
    assert get_codepoint_width(0x1F1FA) == 2  # Regional Indicator U

    # Combining marks
    assert get_codepoint_width(0x0300) == 0  # Combining grave
    assert get_codepoint_width(0x0301) == 0  # Combining acute

    # ZWJ
    assert get_codepoint_width(0x200D) == 0


def test_is_wide_char():
    assert is_wide_char(0x4E00) is True  # CJK
    assert is_wide_char(0xFF21) is True  # Fullwidth A
    assert is_wide_char(ord("A")) is False


def test_is_zero_width():
    assert is_zero_width(0x200D) is True  # ZWJ
    assert is_zero_width(0x0300) is True  # Combining grave
    assert is_zero_width(0x0A) is True  # LF
    assert is_zero_width(ord("A")) is False

File: wasm/test/test_basic.js
import { describe, it } from "mocha";
import assert from "node:assert";
import { getStringWidth, getCodepointWidth, isWideChar, isZeroWidth } from "../index.js";

describe("getStringWidth", () => {
    it("should handle ASCII strings", () => {
        assert.strictEqual(getStringWidth(""), 0);
        assert.strictEqual(getStringWidth("Hello"), 5);
        assert.strictEqual(getStringWidth("Hello, World!"), 13);
    });

    it("should handle CJK strings", () => {
        assert.strictEqual(getStringWidth("中文"), 4);
        assert.strictEqual(getStringWidth("日本語"), 6);
        assert.strictEqual(getStringWidth("한국어"), 6);
    });

    it("should handle mixed strings", () => {
        assert.strictEqual(getStringWidth("Hello, 世界"), 11);
    });

    it("should handle emoji", () => {
        assert.strictEqual(getStringWidth("🐱"), 2);
        assert.strictEqual(getStringWidth("🙈🙉🙊"), 6);
        assert.strictEqual(getStringWidth("👋🏻"), 2);
        assert.strictEqual(getStringWidth("👨‍👩‍👧‍👦"), 2);
        assert.strictEqual(getStringWidth("🇨🇳"), 2);
    });

    it("should handle combining marks", () => {
        assert.strictEqual(getStringWidth("é"), 1);
        assert.strictEqual(getStringWidth("ä"), 1);
        assert.strictEqual(getStringWidth("é̂"), 1);
    });

    it("should handle control characters", () => {
        assert.strictEqual(getStringWidth("hello\n"), 5);
        assert.strictEqual(getStringWidth("hello\tworld"), 10);
    });

    it("should handle fullwidth characters", () => {
        assert.strictEqual(getStringWidth("ＡＢＣ"), 6);
    });
});

describe("getCodepointWidth", () => {
    it("should return 1 for ASCII", () => {
        assert.strictEqual(getCodepointWidth(0x41), 1);  // 'A'
        assert.strictEqual(getCodepointWidth(0x7A), 1);  // 'z'
        assert.strictEqual(getCodepointWidth(0x20), 1);  // space
    });

    it("should return 0 for control characters", () => {
        assert.strictEqual(getCodepointWidth(0x00), 0);  // NULL
        assert.strictEqual(getCodepointWidth(0x0A), 0);  // LF
        assert.strictEqual(getCodepointWidth(0x0D), 0);  // CR
    });

    it("should return 2 for CJK", () => {
        assert.strictEqual(getCodepointWidth(0x4E00), 2);
        assert.strictEqual(getCodepointWidth(0x3042), 2);  // Hiragana A
    });

    it("should return 2 for fullwidth", () => {
        assert.strictEqual(getCodepointWidth(0xFF01), 2);  // Fullwidth !
        assert.strictEqual(getCodepointWidth(0xFF21), 2);  // Fullwidth A
    });

    it("should return 2 for emoji", () => {
        assert.strictEqual(getCodepointWidth(0x1F600), 2);  // Grinning face
        assert.strictEqual(getCodepointWidth(0x1F1FA), 2);  // Regional Indicator U
    });

    it("should return 0 for combining marks", () => {
        assert.strictEqual(getCodepointWidth(0x0300), 0);  // Combining grave
        assert.strictEqual(getCodepointWidth(0x0301), 0);  // Combining acute
    });

    it("should return 0 for ZWJ", () => {
        assert.strictEqual(getCodepointWidth(0x200D), 0);
    });
});

describe("isWideChar", () => {
    it("should return true for CJK", () => {
        assert.strictEqual(isWideChar(0x4E00), true);
    });

    it("should return true for fullwidth", () => {
        assert.strictEqual(isWideChar(0xFF21), true);
    });

    it("should return false for ASCII", () => {
        assert.strictEqual(isWideChar(0x41), false);
    });
});

describe("isZeroWidth", () => {
    it("should return true for ZWJ", () => {
        assert.strictEqual(isZeroWidth(0x200D), true);
    });

    it("should return true for combining marks", () => {
        assert.strictEqual(isZeroWidth(0x0300), true);
    });

    it("should return true for control characters", () => {
        assert.strictEqual(isZeroWidth(0x0A), true);
    });

    it("should return false for ASCII", () => {
        assert.strictEqual(isZeroWidth(0x41), false);
    });
});
